Higher order syntactic and semantic profiler transforms

aishwariyachakraborty opened the pull request:
## Why are these changes needed?
This PR implements the higher order syntactic and semantic profiler transforms as a downstream application of the UBSR transform.

## Related issue number (if any).

aishwariyachakraborty committed:
profiler transforms

Signed-off-by: aishwariyachakraborty <aishwariya.chakraborty@gmail.com>

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/README.md:
````diff
@@ -0,0 +1,64 @@
+the options provided by 
+the [python launcher](../../../../data-processing-lib/doc/python-launcher-options.md).
+```
+  --hosp_metrics_list HOSP_METRICS_LIST
````
Need to be more specific about what are the allowed options and the defaults.  Following the format used [here](https://github.com/IBM/data-prep-kit/tree/dev/transforms/language/pdf2parquet/python#ingest-pdf-to-parquet) would be helpful.

daw3rd commented on transforms/code/semantic_profiler/python/src/sp_transform.py:
```diff
@@ -0,0 +1,124 @@
+        # Make sure that the param name corresponds to the name used in apply_input_params method
+        # of SemanticProfilerTransformConfiguration class
+        super().__init__(config)
+        self.ikb_file = config.get("ikb_file", "../src/ikb/ikb_model.csv")
```
Can you confirm that the wheel built for this transform 1) includes the csv files and 2) that they can be accessed from within the wheel file?  If not, these defaults would only seem to work within the repo, but we need defaults that work when the wheel is used.

aishwariyachakraborty replied:
I have made changes to the pyproject.toml files. The files are present and can be accessed.

aishwariyachakraborty committed:
changes to ensure data files are present in wheel

Signed-off-by: aishwariyachakraborty <aishwariya.chakraborty@gmail.com>

aishwariyachakraborty committed:
test-src working

Signed-off-by: aishwariyachakraborty <aishwariya.chakraborty@gmail.com>

aishwariyachakraborty committed:
Updated params in readme

Signed-off-by: aishwariyachakraborty <aishwariya.chakraborty@gmail.com>

daw3rd commented on .github/workflows/test-code-semantic_profiler.yml:
please merge your branch with dev and regenerate this file.

aishwariyachakraborty replied:
I have made all the changes you have requested. I do not have an option to merge yet. It shows : Merging is blocked
Merging can be performed automatically once the requested changes are addressed.

daw3rd replied:
The merge I'm referring to is one you do locally into your fork.

daw3rd commented on .github/workflows/test-code-higher_order_syntactic_profiler.yml:
please merge your branch with dev and regenerate this file.

daw3rd commented on transforms/code/higher_order_syntactic_profiler/Makefile:
```diff
@@ -0,0 +1,70 @@
+	$(MAKE) RULE=$@ .recurse
+
+.PHONY: workflow-venv
+workflow-venv:
```
Please match these rules with the latest from dev. For example, https://github.com/IBM/data-prep-kit/blob/2a86cec4ddf0bbb0d43c6eb403d4b6c08a7c6a19/transforms/universal/noop/Makefile#L56

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/pyproject.toml:
```diff
@@ -0,0 +1,56 @@
+license = {text = "Apache-2.0"}
+readme = {file = "README.md", content-type = "text/markdown"}
+authors = [
+    { name = "David Wood", email = "dawood@us.ibm.com" },
```
Please provide correct names here.

aishwariyachakraborty replied:
done

daw3rd commented on transforms/code/higher_order_syntactic_profiler/ray/pyproject.toml:
```diff
@@ -0,0 +1,46 @@
+license = {text = "Apache-2.0"}
+readme = {file = "README.md", content-type = "text/markdown"}
+authors = [
+    { name = "David Wood", email = "dawood@us.ibm.com" },
```
Please provide correct names here.

daw3rd commented on transforms/code/semantic_profiler/python/pyproject.toml:
```diff
@@ -0,0 +1,51 @@
+license = {text = "Apache-2.0"}
+readme = {file = "README.md", content-type = "text/markdown"}
+authors = [
+    { name = "David Wood", email = "dawood@us.ibm.com" },
```
Please provide correct names here.

daw3rd commented on transforms/code/semantic_profiler/ray/pyproject.toml:
```diff
@@ -0,0 +1,46 @@
+license = {text = "Apache-2.0"}
+readme = {file = "README.md", content-type = "text/markdown"}
+authors = [
+    { name = "David Wood", email = "dawood@us.ibm.com" },
```
Please provide correct names here.

daw3rd commented on transforms/code/semantic_profiler/Makefile:
```diff
@@ -0,0 +1,70 @@
+	@# Help: Recursively make $@ in all subdirs 
+	$(MAKE) RULE=$@ .recurse
+
+.PHONY: workflow-venv
```
Please match these rules with the latest from dev. For example, https://github.com/IBM/data-prep-kit/blob/2a86cec4ddf0bbb0d43c6eb403d4b6c08a7c6a19/transforms/universal/noop/Makefile#L56

aishwariyachakraborty committed:
Made tha review changes requested
1. Updated makefiles to new the format used in noop
2. Updated author names in toml file

Signed-off-by: aishwariyachakraborty <aishwariya.chakraborty@gmail.com>

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/README.md:
```diff
@@ -0,0 +1,70 @@
+
+| Parameter  | Default  | Description  |
+|------------|----------|--------------|
+| `HOSP_METRICS_LIST`         | `CCR`        | Metrics to be calculated for profiling. Multiple metrics can be entered separated by space. Only valid metric is `CCR` as of now. |
```
I believe the key needs to be in lower case, per the transform implementation.  And, the transform seems to look for `metrics` not `metrics_list` and definitely not `hosp_metrics_list`.

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/pyproject.toml:
```diff
@@ -0,0 +1,55 @@
+    { name = "Aishwariya Chakraborty", email = "aishwariya.chakraborty1@ibm.com" },
+]
+dependencies = [
+    "data-prep-toolkit==0.2.1.dev0",
```
version now needs to be 0.2.2.dev0 everywhere (per the .make.versions in the current dev branch).

aishwariyachakraborty commented:
I have made all the changes you have requested. I do not have an option to merge yet. It shows : Merging is blocked
Merging can be performed automatically once the requested changes are addressed.

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/pyproject.toml:
```diff
@@ -0,0 +1,55 @@
+[project]
+name = "dpk_hosp_transform_python"
+version = "0.2.1.dev0"
```
version now needs to be 0.2.2.dev0 everywhere (per the .make.versions in the current dev branch).

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/README.md:
````diff
@@ -0,0 +1,70 @@
+the options provided by 
+the [python launcher](../../../../data-processing-lib/doc/python-launcher-options.md).
+```
+  --hosp_metrics_list HOSP_METRICS_LIST
````
Can you use https://github.com/IBM/data-prep-kit/blob/dev/transforms/language/pdf2parquet/python/README.md as a template for documenting the configurations and annotations.  In particular, you haven't discussed the annotations that are added, but following the format for the configuration would be good too.

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/src/hosp_transform.py:
```diff
@@ -0,0 +1,167 @@
+        input parquet to the output folder, without modification.
+        """
+        self.logger.debug(f"Transforming one table with {len(table)} rows")
+        if self.metrics_list is not None:
```
You should check for this error/condition in the initializer and throw an exception there if a bad value, such as None.

daw3rd replied:
It appears your are adding a column whose name is the name of the metric.  This is not clear from README.

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/src/hosp_transform.py:
```diff
@@ -0,0 +1,167 @@
+        # Make sure that the param name corresponds to the name used in apply_input_params method
+        # of HigherOrderSyntacticProfilerTransformConfiguration class
+        super().__init__(config)
+        self.metrics_list = config.get("metrics", ["CCR"])
```
"metrics" is not what you want here I think.  Instead, use the `metrics_list` variable, defined above.

aishwariyachakraborty replied:
done, thanks

daw3rd commented on transforms/code/higher_order_syntactic_profiler/python/src/output.html:
This file should probably be removed

daw3rd commented on transforms/code/higher_order_syntactic_profiler/ray/pyproject.toml:
```diff
@@ -0,0 +1,45 @@
+[project]
+name = "dpk_hosp_transform_ray"
+version = "0.2.1.dev0"
```
version number

daw3rd commented on transforms/code/higher_order_syntactic_profiler/ray/pyproject.toml:
```diff
@@ -0,0 +1,45 @@
+    { name = "Aishwariya Chakraborty", email = "aishwariya.chakraborty1@ibm.com" },
+]
+dependencies = [
+    "dpk-hosp-transform-python==0.2.1.dev0",
```
version numbers

daw3rd commented on transforms/code/semantic_profiler/python/README.md:
```diff
@@ -0,0 +1,73 @@
+The set of dictionary keys holding [SPTransform](src/sp_transform.py) 
+configuration for values are as follows:
+
+* _sp_ikb_file_ - This is the path to the IKB file which is a CSV file and by default located in the [IKB](src/ikb/) folder.
```
Please follow this https://github.com/IBM/data-prep-kit/blob/dev/transforms/language/pdf2parquet/python/README.md and distinguish between transform configuration keys and cli parameters (the latter having the `sp_` prefix).

daw3rd replied:
In following the above template, please include the annotations that are added.

daw3rd commented on transforms/code/semantic_profiler/python/README.md:
```diff
@@ -0,0 +1,73 @@
+
+| Parameter  | Default  | Description  |
+|------------|----------|--------------|
+| `SP_IKB_FILE`         | `ikb/ikb_model.csv`        | Path to IKB file. |
```
I don't hink you want caps or `SP_` here.

daw3rd commented on transforms/code/semantic_profiler/python/pyproject.toml:
```diff
@@ -0,0 +1,50 @@
+[project]
+name = "dpk_sp_transform_python"
+version = "0.2.1.dev0"
```
versions

daw3rd commented on transforms/code/semantic_profiler/python/pyproject.toml:
```diff
@@ -0,0 +1,50 @@
+    { name = "Aishwariya Chakraborty", email = "aishwariya.chakraborty1@ibm.com" },
+]
+dependencies = [
+    "data-prep-toolkit==0.2.1.dev0",
```
versions

daw3rd commented on transforms/code/semantic_profiler/python/src/examples/examples-i.csv:
DO these belong in the wheel?  If they are really examples, maybe this examples directory needs to be a sibling of the `src` directory.

aishwariyachakraborty replied:
No these are data files to be used by the code, same as the csv files.

daw3rd commented on transforms/code/semantic_profiler/python/src/ikb/ikb_model.csv:
if you really want to distribute. these, maybe they belong in a `data` directory that is a sibling of `src` and then include them in the wheel via the toml file.  My problem with this is that I'm wondering how someone would specify this file (or other data files) when using the wheel and not the repo project.  If they can be referenced from within the wheel, it would be better to specify `data/ikb/ikb_model.csv` than `src/ikb/ikb_model.csv` I think.

daw3rd commented on transforms/code/semantic_profiler/ray/pyproject.toml:
```diff
@@ -0,0 +1,45 @@
+[project]
+name = "dpk_sp_transform_ray"
+version = "0.2.1.dev0"
```
versions

daw3rd commented on transforms/code/semantic_profiler/ray/pyproject.toml:
```diff
@@ -0,0 +1,45 @@
+    { name = "Aishwariya Chakraborty", email = "aishwariya.chakraborty1@ibm.com" },
+]
+dependencies = [
+    "dpk-sp-transform-python==0.2.1.dev0",
```
versions

daw3rd commented on transforms/code/semantic_profiler/ray/test-data/expected/metadata.json:
```diff
@@ -0,0 +1,46 @@
+    "pipeline": "pipeline_id",
+    "job details": {
+        "job category": "preprocessing",
+        "job name": "NOOP",
```
this file needs to be regenerated since it seems to have been copied from the NOOP project.

touma-I reviewed (changes requested):
@aishwariyachakraborty: Please slack me when you have change to discuss - Internal ID touma@us.ibm.com

aishwariyachakraborty commented:
It was suggested that we combine this with the Syntactic Construct Extractor transform and create a single transform. So I am closing this PR. I will create a new PR after the PR for Syntactic Construct Extractor transform is merged with the dev branch. Hope this is ok?
